### PR TITLE
feat: support query params in list options

### PR DESCRIPTION
**Pull request by wbh1**
## 📝 Description

**What does this PR do and why is this change necessary?**
This will allow for setting custom query parameters in API calls by providing a map.

## ✔️ How to Test

**What are the steps to reproduce the issue or verify the changes?**
Include QueryParams in your ListOptions variable for endpoints that support query parameters.

**How do I run the relevant unit/integration tests?**
N/A

### CODE REVIEW

**Comment by lgarber-akamai** [comment]
@wbh1 How would you feel about an implementation using structs and query tags? 
https://github.com/linode/linodego/commit/43bc40244851bc45bf9fbbf7e31ce91b0b605c93

**Comment by zliang-akamai** [comment]
@lgarber-akamai Is this PR ready yet?

**Review by jriddle-linode — APPROVED** [review]

File: pagination.go [hunk 1]
```diff
@@ -102,3 +124,60 @@ func (c *Client) listHelper(ctx context.Context, pager PagedResponse, opts *List
...
+
+// flattenQueryStruct flattens a structure into a Resty-compatible query param map.
+// Fields are mapped using the `query` struct tag.
+func flattenQueryStruct(val any) (map[string]string, error) {
+	result := make(map[string]string)
+
+	reflectVal := reflect.ValueOf(val)
+
+	// Deref pointer if necessary
+	if reflectVal.Kind() == reflect.Ptr {
```
Comment (zliang-akamai) [inline pagination.go:136]:
```suggestion
	if reflectVal.Kind() == reflect.Pointer {
```
[Ptr is the old name for the Pointer kind. ](https://pkg.go.dev/reflect#pkg-constants)

File: pagination.go [hunk 2]
```diff
@@ -102,3 +124,60 @@ func (c *Client) listHelper(ctx context.Context, pager PagedResponse, opts *List
...
+		reflectVal = reflect.Indirect(reflectVal)
```
Comment (zliang-akamai) [inline pagination.go:137]: Can we handle `nil` pointer here?

File: pagination.go [hunk 3]
```diff
@@ -102,3 +124,60 @@ func (c *Client) listHelper(ctx context.Context, pager PagedResponse, opts *List
...
+	}
+
+	valType := reflectVal.Type()
```
Comment (zliang-akamai) [inline pagination.go:140]:
```suggestion

	if reflectVal.Kind() != reflect.Struct {
		return result, fmt.Errorf(
			"expected struct type for the QueryParams but got: %s",
			reflectVal.Kind().String(),
		)
	}

	valType := reflectVal.Type()
```

File: pagination.go [hunk 4]
```diff
@@ -102,3 +124,60 @@ func (c *Client) listHelper(ctx context.Context, pager PagedResponse, opts *List
...
+			return nil, fmt.Errorf("invalid query param tag: %s", currentField.Name)
+		}
+
+		// Skip if it's a zero value
+		if valField.IsZero() {
+			continue
+		}
+
+		// Deref the pointer is necessary
+		if valField.Kind() == reflect.Ptr {
```
Comment (zliang-akamai) [inline pagination.go:162]:
```suggestion
		if valField.Kind() == reflect.Pointer {
```

File: pagination.go [hunk 5]
```diff
@@ -102,3 +124,60 @@ func (c *Client) listHelper(ctx context.Context, pager PagedResponse, opts *List
...
+		switch valField.Interface().(type) {
+		case string:
+			resultField = valField.String()
+		case int64, int:
+			resultField = strconv.FormatInt(valField.Int(), 10)
+		case bool:
+			resultField = strconv.FormatBool(valField.Bool())
+		default:
+			return nil, fmt.Errorf("unsupported query param type: %s", valField.Type().Name())
+		}
```
Comment (zliang-akamai) [inline pagination.go:177]: Do we want support slice (array) in query parameters? Maybe not today?

Comment (lgarber-akamai) [reply pagination.go:177]: I don't think it's necessary in this case since query params generally aren't expected to be provided in list format 👍

**Comment by lgarber-akamai** [comment]
@zliang-akamai Thanks for the suggestions!

**Comment by zliang-akamai** [comment]
@lgarber-akamai Are we good to merge this?

File: pagination.go (hunk 5 above)
Comment (zliang-akamai) [inline pagination.go]: We may want to switch to reflect types, like we did in TF filter.

**Review by zliang-akamai — CHANGES_REQUESTED** [review]
I can make this change later today

**Review by zliang-akamai — APPROVED** [review]

**Review by lgarber-akamai — APPROVED** [review]

**Review by zliang-akamai — APPROVED** [review]

**Pull request merged**